chore: unbiased and un-opionated assignment changes (gigantic) (#6)
seenons/seenons-be-assignment: README.md

- Service provider: Must corresponde to a prefedined record in a list of service providers.

### Specific Use Case Requirements

1. **Service Partners Availability**
   Ensure the date for waste pickup is not before tomorrow, accounting for a next-day service policy.
   see: <https://github.com/seenons/seenons-be-assignment/blob/main/src/usecases/availability/find_availability_at.ts>

2. **Register Waste Pickup**
   The selected date must be within the service provider's scheduling constaints and waste stream availability.
   see: <https://github.com/seenons/seenons-be-assignment/blob/main/src/usecases/pickup/register_waste_pickup.ts>

Notes:

- You can use the provided code as a starting point or rebuild it however you see fit, including the domain representations in case you do not agree with its current contents.

## The delivery

---

- Create a **private** copy of this repository. Do not fork it.
- Create a new readme file that explains your take on the assignment, your thought process, and any other information that you believe is relevant for the review.
- Implement the necessary changes to at least achieve the requirements.

> - Invite <caio@seenons.com> as a collaborator to your private repository.

Reviewer: I don't know if Github allows this, but can this be `@seenons/backenders` instead? You do the reviews now, but that might not always be the case. It's better if everybody has access.

Author: Sadly no, we could use a bot (perhaps wall-e for that)

Hello there!

At Seenons, we are building a platform to support a circular economy. This encompasses collection, transportation and re-purposing of different waste streams and our assignment is a subset of what we do.

## The assignment

---

The task is to at least complete two use cases and their corresponding specs.

- Retrieve availability of service providers at a given location, waste stream and date.
  - see: <https://github.com/seenons/seenons-be-assignment/blob/main/src/usecases/availability/find_availability_at.ts>
- Register a waste stream pickup to be performed by a service provider.
  - see: <https://github.com/seenons/seenons-be-assignment/blob/main/src/usecases/pickup/register_waste_pickup.ts>

You can use the provided code as a starting point or rebuild it however you see fit, including the domain representations in case you do not agree with its contents.

## The delivery

---

> - Create a copy of this repository, do not fork it.

Reviewer 2: Maybe be specific about this, since you mention private in the last bullet point
```suggestion
- Create a **private** copy of this repository. Do not fork it.
```

### Implementation (required)

Waste Stream Pickup Registration

- Make sure that a Stream can be picked up by the Service Provider at the given date

Service Providers Availability

- Make sure that when searching for availability, the following lines are respected per postal code and date:
- You can use the provided data set to achieve it. see: <https://github.com/seenons/seenons-be-assignment/blob/main/src/domain/dataset/service_providers.data.ts>

| postal_code |    date    |                      result                      |
| :---------: | :--------: | :----------------------------------------------: |
|    1010     | 2023-10-02 | [Unwasted (paper, metal), Bluecollection(metal)] |
|    1010     | 2023-10-04 |    [Unwasted(metal) , Bluecollection(metal)]     |
|    2000     | 2023-10-05 |             [Bluecollection(metal)]              |
|    1010     | 2023-10-08 |                        []                        |
|    0000     | 2023-10-03 |                        []                        |

### Test Coverage (required)

- Complete Register Waste Pickup Spec.
- Complete Find Availability Spec.

### Bonus Points (not required but a nice to have)

#### Opportunities

> - Implement a Database Provider and whatever layer of communication you deem neccessary?

Reviewer 2: ```suggestion
- Implement a Database Provider and whatever layer of communication you deem neccessary.
```

- npm
- TypeScript

Although experience with these technologies is not a requirement to join our team, we are looking for people who will be
able to work in this ecosystem, so we recommend you to use them.

If you are not familiar with these technologies, you can also complete the assignment using the ecosystems below.

- .NET (Core) 5+
- Java
- Kotlin
- Python

Would you rather do it in another language? Let us know beforehand!

## Disclaimer

---

Please note that this assignment is not a direct representation of how Seenons built its software.

The assignment is meant for all levels of seniority in mind, so the concepts, patterns, domain models and tools have been simplified to ensure fairness.

## FAQ

- Do I have to use the base project for my assignment? No
- Do I have to use exactly the payload and response models that were provided? No
- Can I rename or move the files? Yes
- Can I use architectural patterns X or Y? Please do.
- Do I have to use a database? No, you can use in memory objects or the provided data set
- Can I expose this as an API instead? Yes
- Can I do more than asked? Yes, the time is yours
- Is there a deadline? No, but please be mindful that its FIFO.

> - How long does it take to complete the assignment? You can use as much time as you want, make it nice, make it useful, make it print ascii art, you choose as its your personal time.

Reviewer 2: ```suggestion
- How long does it take to complete the assignment? You can use as much time as you want. Make it nice, make it useful, or make it print ASCII art. You choose, as it's your personal time.
```
